analysis/rekognition_worker.py

```diff
@@ -0,0 +1,116 @@
+import logging as log
+import time
+import concurrent.futures
+import worker.settings as settings
+from functools import partial
+from collections import defaultdict, Counter
+from confluent_kafka import Consumer, Producer
+from analysis.task import handle_image_task, TaskStatus
+from analysis.util import LocalTokenBucket, RecentlyProcessed, parse_msg
+
+"""
+Feeds crawled images stored in S3 to AWS Rekognition and stores the resulting
+metadata in a new Kafka topic.
+"""
+
+IMG_BUCKET = 'cc-image-analysis'
+# Used to respect AWS service limits
+MAX_REKOGNITION_RPS = 50
+# Number of pending messages to store simultaneously in memory
+NUM_MESSAGES_BUFFER = 1500
+MAX_PENDING_FUTURES = 1000
+# These are network-bound threads, it's OK to use a lot of them
+NUM_THREADS = 50
+# Number of recently processed image IDs to retain for duplication prevention
+NUM_RECENT_IMAGE_ID_RETENTION = 10000
+
+
+def _monitor_futures(futures):
+    """
+    Summarizes task progress and handles exceptions. Returns a list of pending
+    futures (with completed futures removed).
+    """
+    _futures = []
+    statuses = defaultdict(int)
+    for f in futures:
+        if f.done():
+            try:
+                res = f.result()
+                statuses[res] += 1
+            except botocore.exceptions.ClientError:
+                log.warning("Boto3 failure: ", exc_info=True)
+                statuses[TaskStatus.ERROR] += 1
+        else:
+            # Preserve pending futures
+            _futures.append(f)
+    return _futures, Counter(statuses)
+
+
+def _poll_work(msg_buffer, msgs_remaining, consumer):
+    """ Poll consumer for messages and parse them. """
+    while len(msg_buffer) < NUM_MESSAGES_BUFFER and msgs_remaining:
+        msg = consumer.poll(timeout=10)
+        if not msg:
+            log.info('No more messages remaining')
+            msgs_remaining = False
+        else:
+            parsed = parse_msg(msg)
+            if parsed:
+                msg_buffer.append(parsed)
+    return msgs_remaining
+
+
+def _schedule_tasks(msg_buffer, executor, futures, task_fn, producer):
+    token_bucket = LocalTokenBucket(MAX_REKOGNITION_RPS)
+    recent_ids = RecentlyProcessed(NUM_RECENT_IMAGE_ID_RETENTION)
+    if len(futures) < MAX_PENDING_FUTURES:
+        for msg in msg_buffer:
+            partial_task = partial(task_fn, msg, producer, recent_ids)
+            future = executor.submit(token_bucket.throttle_fn, partial_task)
+            futures.append(future)
+        # Flush msg buffer after scheduling
+        msg_buffer = []
+    return msg_buffer
+
+
+def listen(consumer, producer, task_fn):
+    status_tracker = Counter({})
+    msg_buffer = []
+    futures = []
+    executor = concurrent.futures.ThreadPoolExecutor(max_workers=NUM_THREADS)
+    msgs_remaining = True
+    last_log = None
+    while msgs_remaining:
+        msgs_remaining = _poll_work(msg_buffer, msgs_remaining, consumer)
+        msg_buffer = _schedule_tasks(
```

Comment:
For me, having the `msg_buffer` mutated in place by `_poll_work` but not returned, and then mutated _and_ returned by `_schedule_tasks` makes it a bit tricky to follow the logic (for example, it took a bit to figure out where the first items are added to the buffer).  My preference would be pretending the lists are immutable from outside the functions (i.e,. always returning them) but I think that might not be optimal performance-wise.

Reply:
You confirmed my suspicions that that section is a little bit smelly. I rewrote it to reduce the amount of mutation of the message buffer inside of the functions. I still don't like that both polling and scheduling have to modify the message buffer but think it's the lesser evil compared to joining these two functions together again.